### ISSUE TITLE
[no-await-in-loop] should check for `await` keyword inside `forEach` loop

### ISSUE THREAD

**Issue by dimabory** [eslint/eslint#12576]
**What rule do you want to change?**

[**no-await-in-loop**
](https://eslint.org/docs/rules/no-await-in-loop#disallow-await-inside-of-loops-no-await-in-loop)

**Does this change cause the rule to produce more or fewer warnings?**
More.

**How will the change be implemented? (New option, new default behavior, etc.)?**
It supposed to include new checks.

**Please provide some example code that this change will affect:**
```js
function delay() {
  return new Promise(resolve => setTimeout(resolve, 300));
}

async function delayedLog(item) {
  await delay();
  console.log(item);
}
async function processArray(array) {
  array.forEach(async (item) => {
    await delayedLog(item);
  })
  console.log('Done!');
}

processArray([1, 2, 3]);
```
Output:
```
Done!
1
2
3
```
---
Link to read https://lavrton.com/javascript-loops-how-to-handle-async-await-6252dd3c795/

**What does the rule currently do for this code?**
It allows using `await` inside `forEach` loop.

**What will the rule do after it's changed?**
It will warn "_Unexpected `await` inside a loop._". Probably it makes sense to warn a developer kind used of the loop. (for/of, forEach, etc.).

**Are you willing to submit a pull request to implement this change?**
Yes, it's a good chance to start contributing `eslint`.

**Comment by kaicataldo** [comment]
Thanks for the proposal. Can you explain what problem you’re trying to solve? If the intended behavior of the example code is to log “Done!” after the array values have resolved, there’s definitely a bug in this code. However, given that the rule is intended to help catch cases where async code is accidentally run in serial rather in parallel and since the promises in this example are being awaited in parallel, I wouldn’t expect it to warn on this code.

**Comment by dimabory** [comment]
Hi, @kaicataldo 🙂

I got your point. It seems like I provided a controversial example, which can be either correct or not depending on the developer's needs.

I would like to give another one where an async code actually supposed to be run in parallel:
```js
async function foo(v) {
 return new Promise(resolve => setTimeout(() => resolve(v ** 2), 1));
}

async function bar(array) {
 const result = [];
 array.forEach(async (i) => result.push(await foo(i)));
 return result;
}

await bar([1,2,3,4,5,6]).length === 0; // true

```

At first sight, it looks like the expected output is a fulfilled list of values, but the actual return value is an empty list.

How it should be implemented?
```js
async function bar(array) {
 const r = [];
 array.forEach((i) => r.push(foo(i)));
 return Promise.all(r);
}

await bar([1,2,3,4,5,6]).length === 0; // false

**Comment by kaicataldo** [comment]
Agreed that this is buggy code, but I again don’t think it’s this rule’s responsibility to catch this, since the problem isn’t the use of `Array.prototype.forEach`, but the lack of awaiting the resolution of all the promises in the array.

While I do think catching this would be useful, I don’t think it’s the responsibility of this rule and I’m not sure how we could do it reliably without type information. 

Do you have an idea how we could enforce this reliably? Please note that ESLint only runs on one file at a time.

**Comment by dimabory** [comment]
> ... _the problem isn’t the use of `Array.prototype.forEach`, but the lack of awaiting the resolution of all the promises in the array._ 
> ... _I don’t think it’s the responsibility of this rule_ 

Hm... I'm just curious about the difference between the example provided for [no-await-rule](https://eslint.org/docs/rules/no-await-in-loop#disallow-await-inside-of-loops-no-await-in-loop) rule using for/of statement vs. `Array.prototype.forEach` :
```js
// https://eslint.org/docs/rules/no-await-in-loop#disallow-await-inside-of-loops-no-await-in-loop
async function foo(things) {
  const results = [];
  for (const thing of things) {
    // Bad: each loop iteration is delayed until the entire asynchronous operation completes
    results.push(await bar(thing));
  }
  return baz(results);
}
// example with Array.prototype.forEach
async function foo(things) {
 const results = [];
 things.forEach(async thing => results.push(await bar(thing)));
 return baz(results);
}
```

**Comment by dimabory** [comment]
@kaicataldo I've been thinking how to deal with such kind of buggy code, but TBH I have no idea right now how to enforce this rule 🤷‍♂️ ... and yes, you're definitely right that it has little to do with _no-await-in-loop_ rule.
